spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/DeleteStep.java

```diff
@@ -0,0 +1,88 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.skipper.server.deployer.strategies;
+
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.cloud.deployer.spi.app.AppDeployer;
+import org.springframework.cloud.deployer.spi.app.AppStatus;
+import org.springframework.cloud.deployer.spi.app.DeploymentState;
+import org.springframework.cloud.skipper.domain.Release;
+import org.springframework.cloud.skipper.domain.Status;
+import org.springframework.cloud.skipper.domain.StatusCode;
+import org.springframework.cloud.skipper.server.domain.AppDeployerData;
+import org.springframework.cloud.skipper.server.repository.AppDeployerDataRepository;
+import org.springframework.cloud.skipper.server.repository.DeployerRepository;
+import org.springframework.cloud.skipper.server.repository.ReleaseRepository;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Responsible for deleting the provided list of applications and updating the status of
+ * the release.
+ * @author Mark Pollack
+ */
+public class DeleteStep {
+
+	private final Logger logger = LoggerFactory.getLogger(DeleteStep.class);
+
+	private final ReleaseRepository releaseRepository;
+
+	private final DeployerRepository deployerRepository;
+
+	private final AppDeployerDataRepository appDeployerDataRepository;
+
+	public DeleteStep(ReleaseRepository releaseRepository, DeployerRepository deployerRepository,
+			AppDeployerDataRepository appDeployerDataRepository) {
+		this.releaseRepository = releaseRepository;
+		this.deployerRepository = deployerRepository;
+		this.appDeployerDataRepository = appDeployerDataRepository;
+	}
+
+	@Transactional
+	public Release delete(Release release, AppDeployerData existingAppDeployerData,
+			List<String> applicationNamesToDelete) {
+
+		AppDeployer appDeployer = this.deployerRepository.findByNameRequired(release.getPlatformName())
+				.getAppDeployer();
+
+		Map<String, String> appNamesAndDeploymentIds = existingAppDeployerData.getDeploymentDataAsMap();
+
+		for (Map.Entry<String, String> appNameAndDeploymentId : appNamesAndDeploymentIds.entrySet()) {
+			if (applicationNamesToDelete.contains(appNameAndDeploymentId.getKey())) {
+				AppStatus appStatus = appDeployer.status(appNameAndDeploymentId.getValue());
+				if (appStatus.getState().equals(DeploymentState.deployed)) {
+					appDeployer.undeploy(appNameAndDeploymentId.getValue());
+				}
+				else {
+					logger.warn("For Release name {}, did not undeploy existing app {} as it status is not 'deployed'.",
+							release.getName(),
+							appNameAndDeploymentId.getKey());
+				}
+			}
+		}
+
+		Status deletedStatus = new Status();
+		deletedStatus.setStatusCode(StatusCode.DELETED);
```

Comment:
We might be marking the status to `DELETED` even for those who didn't have `deployed` state (the ones handled in else block above) ?

Reply:
The way I interpret the status Code is that those apps for the release have been requested to be deleted.  The ones that carry over between releases done count.  We asked 'the platform' to delete the release.  That may have been one app or many, but the status of the release is 'requested deleted'. Does that make sense?

Reply:
yeah, it would make sense from the release point of view. But, once the release is marked `DELETED`, there could be subsequent releases to re-install them. In case where the platform didn't handle the deletion successfully, then the subsequent re-install operations would have apps from completely different package etc. based on the release's install properties.

Reply:
If deleted fail (which I've actually never seen) there would still be running apps.  Any subsequent install operations for that release name would install whatever was requested.  In terms of it being completely different apps/packages - that problem exists in the sunny day scenario as well (delete succeedes).  The 'tis why I think there was the option in helm to by default not to be able to install and app that was deleted - I created an issue for us to investigate that more.

Reply:
https://github.com/spring-cloud/spring-cloud-skipper/issues/207

Reply:
The behavior was the same when this was synchronous...

Reply:
Right, this issue irrespective of sync vs async as it is all about what status we want to update the apps when they are not in already deployed state but deletion is requested.

Reply:
Right.  Will change to that if we fail to get the health within the timeout and in the general case of failure, we delete the apps we asked to deploy for the new release.  Status will be set to failed.  Will prob then need to change the 'findLatestRelease' to be findLatestReleaseNotFailedNotUnknown...